esp/logging/logginglib/modularlogagent.cpp

```diff
@@ -0,0 +1,1526 @@
+{
+}
+
+inline bool delegateUpdateLog(IUpdateLog& delegate, const char* content, IEspUpdateLogResponse& response)
+{
+    delegate.updateLog(content, response);
+    return (response.getStatusCode() == 0);
+}
+
+void CDelegatingAgent::updateLog(IEspUpdateLogRequestWrap& request, IEspUpdateLogResponse& response)
+{
+    bool result = false;
+    if (!m_updateLog)
+        ierrlog(Major, "UpdateLog service module does not exist");
+    else if (!m_updateLog->isEnabled())
+        m_updateLog->updateLog(nullptr, response);
+    else
+    {
+        const char* updateLogRequest = request.getUpdateLogRequest();
+        if (!isEmptyString(updateLogRequest))
+        {
+            result = delegateUpdateLog(*m_updateLog, updateLogRequest, response);
+        }
+        else
+        {
+            IPTree* ulrTree = request.getLogRequestTree();
+            if (ulrTree)
+            {
+                StringBuffer xml;
+                toXML(ulrTree, xml);
+                result = delegateUpdateLog(*m_updateLog, xml, response);
+            }
+            else
+            {
+                m_updateLog->ierrlog(Major, "requires XML markup; none given");
+            }
+        }
+    }
+
+    if (result && response.getStatusCode() != 0)
+    {
+        response.setStatusCode(0);
+    }
+    else if (!result && response.getStatusCode() == 0)
+    {
+        response.setStatusCode(-1);
+    }
+}
+
+bool CDelegatingAgent::hasService(LOGServiceType type) const
+{
+    switch (type)
+    {
+    case LGSTGetTransactionSeed: return false;
+    case LGSTGetTransactionID:   return false;
+    case LGSTUpdateLOG:          return m_updateLog && m_updateLog->isEnabled();
+    case LGSTterm:
+        ierrlog(Major, "invalid service request for LGSTterm");
+        break;
+    default:
+        ierrlog(Major, "unexpected service request for %d", int(type));
+        break;
+    }
+    return false;
+}
+
+bool CDelegatingAgent::configureUpdateLog(const IPTree& configuration, const CModuleFactory& factory)
+{
+    static Owned<IPTree> defaultConfiguration(createPTreeFromXMLString(VStringBuffer("<%s><%s/></%s>", moduleAgent, moduleUpdateLog, moduleAgent)));
+    Linked<const IPTree> effectiveConfiguration;
+    const char* updateLogValue = configuration.queryProp(propUpdateLog);
+    bool hasScalar = !isEmptyString(updateLogValue);
+    bool hasObject = configuration.hasProp(moduleUpdateLog);
+    if (hasScalar && hasObject)
+    {
+        uerrlog(Major, "%s and %s cannot be configured at the same time", propUpdateLog, moduleUpdateLog);
+        return false;
+    }
+    if (hasScalar)
+    {
+        TokenDeserializer deserializer;
+        bool applyDefault = false;
+        if (deserializer(updateLogValue, applyDefault) != Deserialization_SUCCESS)
+        {
+            uerrlog(Major, "expected Boolean value for %s; found '%s'", propUpdateLog, updateLogValue);
+            return false;
+        }
+        if (!applyDefault)
+            return true;
```

Comment:
If we're not applying the default, what is the actual configuration used in this case?

Reply:
The actual configuration used is worthless. Since you can't a scalar and object and a scalar was given with false, the actual configuration is a delegating agent with no delegates. It's no different than omitting both the scalar and object, or including the object and disabling it.

Reply:
ok